test/integration/framework/common.go

```diff
@@ -0,0 +1,172 @@
+)
+
+func newGardenerTest(kubeconfig string) (*GardenerTest, error) {
+	var (
+		errorResult    error
+		gardenerClient = clientcmd.NewNonInteractiveDeferredLoadingClientConfig(
+			&clientcmd.ClientConfigLoadingRules{ExplicitPath: kubeconfig},
+			&clientcmd.ConfigOverrides{},
+		)
+	)
+
+	k8sGardenClient, err := kubernetes.NewClientFromFile(kubeconfig, nil, client.Options{})
+	if err != nil {
+		errorResult = multierror.Append(errorResult, err)
+	}
+
+	// Create a GardenV1beta1Client and the respective API group testScheme for the Garden API group.
+	gardenerClientConfig, err := kubernetes.CreateRESTConfig(gardenerClient, nil)
+	if err != nil {
+		errorResult = multierror.Append(errorResult, err)
+	}
+
+	gardenerClientSet, err := gardenclientset.NewForConfig(gardenerClientConfig)
+	if err != nil {
+		errorResult = multierror.Append(errorResult, err)
+	}
+
+	k8sGardenClient.SetGarden(gardenerClientSet)
+
+	gardenerTest := GardenerTest{
+		K8sGardenClient:     k8sGardenClient,
+		K8sGardenInformers:  gardeninformers.NewSharedInformerFactory(k8sGardenClient.Garden(), 0),
+		KubeInformerFactory: kubeinformers.NewSharedInformerFactory(k8sGardenClient.Kubernetes(), 0),
+	}
+
+	return &gardenerTest, errorResult
+}
+
+// CreateShootTestArtifacts creates the necessary artifacts for a shoot tests including a random integration test name and
+// a shoot object which is read from the resources directory
+func CreateShootTestArtifacts(shootTestYamlPath string) (string, *v1beta1.Shoot, error) {
+	var errorResult error
+
+	randomString, err := utils.GenerateRandomString(10)
+	if err != nil {
+		return "", nil, err
+	}
+
+	integrationTestName := "test-" + strings.ToLower(randomString)
+
+	shootInterface, err := loadObjectFromSpec(shootTestYamlPath, Shoot)
+	if err != nil {
+		return "", nil, err
+	}
+
+	shootToReturn, ok := shootInterface.(*v1beta1.Shoot)
+	if !ok {
+		if err != nil {
+			return "", nil, fmt.Errorf("failed to convert loaded object to a shoot obect")
+		}
+	}
+
+	shootToReturn.Name = integrationTestName
+	shootToReturnDomain := integrationTestName + ".shoot.dev.k8s-hana.ondemand.com"
+	shootToReturn.Spec.DNS.Domain = &shootToReturnDomain
+
+	return integrationTestName, shootToReturn, errorResult
+
+}
+
+func hasAnnotation(mapToCheck map[string]string, keyToCheck string) bool {
+	for k := range mapToCheck {
+		if k == keyToCheck {
+			return true
+		}
+	}
+	return false
+}
+
+// loadObjectFromSpec loads the contents of file and decodes it as a
+// ControllerManagerConfiguration object.
+func loadObjectFromSpec(file string, fileType Type) (interface{}, error) {
+	if err := garden.AddToScheme(testScheme); err != nil {
+		return nil, err
+	}
+	if err := v1beta1.AddToScheme(testScheme); err != nil {
+		return nil, err
+	}
+	if err := v1beta1.RegisterConversions(testScheme); err != nil {
+		return nil, err
+	}
+
+	data, err := ioutil.ReadFile(file)
+	if err != nil {
+		return nil, err
+	}
+
+	switch fileType {
```

Comment:
Okay, I get your point, I was doing that for clean error handling, otherwise we panic from Go and won't be able to view the error nicely in Gingko style. Plus, this is a test, CreateTestShoot is called once every test anyways. Not really inefficient as you said! 

Regarding the switch, the clusterType needs to be there for you to be able to say `var shoot v1beta1.Shoot` or `var shoot v1beta1.Seed` which is then passed to the independent `Decode`